hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFileBlock.java

```diff
@@ -1721,14 +1765,36 @@ protected HFileBlock readBlockDataInternal(FSDataInputStream is, long offset,
         if (headerBuf == null) {
           headerBuf = onDiskBlock.duplicate().position(0).limit(hdrSize);
         }
-        // Do a few checks before we go instantiate HFileBlock.
-        assert onDiskSizeWithHeader > this.hdrSize;
-        verifyOnDiskSizeMatchesHeader(onDiskSizeWithHeader, headerBuf, offset, checksumSupport);
+
         ByteBuff curBlock = onDiskBlock.duplicate().position(0).limit(onDiskSizeWithHeader);
         // Verify checksum of the data before using it for building HFileBlock.
         if (verifyChecksum && !validateChecksum(offset, curBlock, hdrSize)) {
+          invalidateNextBlockHeader();
+          span.addEvent("Falling back to HDFS checksumming.", attributesBuilder.build());
           return null;
         }
+
+        // TODO: is this check necessary or can we proceed with a provided value regardless of
```

Comment:
so this is to validate that if a caller passes in a size, it matches what we read. i suppose this is good, but maybe depends on how much we trust the caller? not sure where the passed in value comes from at the moment

Reply:
Tracing the call sites, I see four sources of this parameter.

1. Caller doesn't know how big the block is, so it provides `-1`.
2. Caller is seeking to this block via a block index reference, provides an expected block size based on the value that was written to the index. I think those reads are covered by checksum validation, we we can trust this source.
3. In the previous call, the `readBlockDataInternal` method over-read the next block's header and takes a small detour to [read out](https://github.com/apache/hbase/blob/4c1fd0d/hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFileBlock.java#L1719-L1720) and save off the value as a local variable. This value is not validated after the buffer read. Looks like something else that I should invalidate in `invalidateNextBlockHeader()`. We should be skeptical of values with this origin story.
4. Call has read the previous block out of blockcache and so has a hint for the next block's size as provided by the cached previous block. This seems to only happen when we're working with an off-heap BlockCache, like BucketCache or Memcached. See references to `HFileBlock.BLOCK_METADATA_SPACE` and the `Cacheable` interface. The value seems to originate from the same next-block-header over-read that populates (3), and so we should be skeptical of these values as well.

So in today's code, I'd say we don't trust this value very often -- what percentage of block reads are directly driven by a reference from the block index? Maybe after this patch, with more explicit care to invalidate the next header buffer when things look bad, and if we can find a way to mark the value's origin as the checksum-validated read of the block index, probably we trust it a lot, though it'll be absent more often. As of now, I think the common case is the value comes from the pre-read of the next block's header, which means we're just comparing two values that have identical origin and it's a wast of CPU.

Also, we rely on this un-verified value in order to perform checksum validation. Read down into [ChecksumUtil](https://github.com/apache/hbase/blob/1aea663c6de4c08f0b2a2d4b2ca788772dc0b686/hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/ChecksumUtil.java#L172-L178), it looks like we again read values from the unverified block header in order to perform checksum validation over data that includes the header value we're referencing.

Reply:
Thanks for tracing that. So I think 4 actually seems reasonably concerning. However, now I'm trying to think how much this even matters:

- So here we're yet again validating onDiskSizeWithHeader.
- However, if we've reached this point, we've already used onDiskSizeWithHeader to successfully allocate a buffer, read into it, and verify checksums
- So at this point, if verifyChecksum is true then I think we can trust onDiskSizeWithHeader. Otherwise some part of the above surely would have failed.
- In terms of the case where verifyChecksum is false, I suppose it depends on how much we trust HDFS checksums and/or want to validate against a non-corruption hbase-internal bug.

Given the hotness of this codepath, maybe we don't want to do the work of unit tests here (protecting against hbase internal bugs)? So maybe it depends on trusting hdfs checksums, which seems reasonable? In that case, maybe don't do this verification?

WDYT?

Reply:
Another observation: this is after the data in the block header has passed checksum validation (if it was enabled), so presumably the value read from the header is as trustworthy as it'll ever be.

Reply:
Yep, that's the more succinct summary of what I was getting at above :D So I'm thinking we get rid of the check.

Reply:
> Yep, that's the more succinct summary of what I was getting at above :D

Err, sorry. Github didn't refresh to show me your more recent comment.

Reply:
We have a test explicitly checking for this verification. I don't know how realistic the scenario is.

https://github.com/apache/hbase/blob/8f7801c1f11c830fa236fcb555c1cf9b44fe0f92/hbase-server/src/test/java/org/apache/hadoop/hbase/io/hfile/TestHFileBlock.java#L402-L406

Reply:
In my initial review here I didn't realize that the code you had added in this block was basically the replacement for verifyOnDiskSizeMatchesHeader. I suppose we should keep this check despite our observations above, and your changes to it look good. 

If we see this validation failure in the future, I think it means we missed some pathway to validating the original onDiskSizeWithHeader. Even if the data we read is good, might make sense to try to fix. It could also help protect those people who have checksums disabled for whatever reason.